### ISSUE TITLE
Infinite record accepted by the compiler

### ISSUE THREAD

**Issue by FPtje** [elm-lang/elm-compiler#1333]
Take the following piece of code:
```elm
import Html exposing (text)

main =
  text (foo.x)
  
foo : { b | x : String } -- signature can be omitted without affecting the issue
foo = { foo | x = "what" }
```

`foo` is defined by taking `foo` and setting its member `x` to `"what"`. Conceptually, there's something infinite going on here, because `foo` refers to itself. The compiler doesn't mind this, and regards this code as OK. 

The biggest worry with this code in the end is that the main prints `undefined`. This means that `foo.x` is an `undefined` of type `String`. This can cause runtime errors that may or may not be tremendously difficult to debug.

Another property of these infinite records is that the compiler accepts *any* field accessor as valid, having type `a`:

```elm
foo : { b | x : String } -- signature can still be omitted without affecting the issue
foo = { foo | x = "wut" }


a : Bool -- Let's just say this is a bool. It won't matter.
a = foo.fooDoesntHaveThisField -- No compilation errors here
```
With this hack, we can create an `undefined` similar to the one Haskell has:

```elm
import Color exposing (..)
import Graphics.Collage exposing (..)

u = {u | a = ""}

myFancyColor = rgb u.undefined u.undefined u.undefined

main = 
  collage 400 400
    [ filled myFancyColor (circle 110)
    ]
```

The compiler accepts this. Javascript, as error tolerant as it is, thinks it's fine too. The result is a black circle. Even Javascript has its limits though:

```elm
import Color exposing (..)
import Graphics.Collage exposing (..)

u = {u | a = ""}

main = 
  collage 400 400
    [ u.undefined (rgb 0 255 255) (circle 110)
    ]

```

Throws a runtime error: `Cannot read property 'arity' of undefined`.

Basically, all the type safety of Elm can be "disabled" and traded for arbitrary runtime behaviour. Some developers might be crazy enough to actually abuse this to get around the limitations of the type system. This is at least a little bit undesirable.

**Comment by jvoigtlaender** [comment]
The recursion stuff is a duplicate of #873.

The other stuff, I fear, is a misunderstanding on your part of polymorphic types. It's absolutely okay to access arbitrary fields if the value was declared to have arbitrary fields, which is exactly what happens when you declare it to have type { b | ... }.

**Comment by jvoigtlaender** [comment]
Rereading, maybe you are not actually surprised by being able to access an arbitrary field from somethign declared of type `{ b | ... }`, but by that declaration being accepted for the equation you gave.

Still, nothing new here that is not already an issue of #873.

You get the same with:
```elm
x : a
x = x

y : Bool
y = x.foo
```
so it has nothing to do with infinite records specifically.

Fixing #873 will deal with this issue here, so I think this one here can be closed.

**Comment by jvoigtlaender** [comment]
Or, to really make clear it has nothing to do with records specifically:
```elm
x : a
x = x

y : Bool
y = x 12
```
which is also accepted by the compiler right now.

**Comment by FPtje** [comment]
> The other stuff, I fear, is a misunderstanding on your part of polymorphic types. It's absolutely okay to access arbitrary fields if the value was declared to have arbitrary fields, which is exactly what happens when you declare it to have type { b | ... }.

I disagree with you there. Elm 0.16 disables adding and removing fields. With the assumption that you cannot add fields to records, the compiler should know that I cannot just access any arbitrary field. In fact it does know in normal circumstances.

Take a look at this example:

```elm
a = {b = "asdf"}

c : { r | b : String } -> { r | b : String }
c r = {r | b = "bla"}

d = (c a).d
```
This gives a type mismatch, saying that `(c a)` doesn't have a field `d`. This is despite the fact that c returns a polymorphic record of type `{ r | b : String }`.
On the other hand, if I give it a record that *does* have `d`, it works:

```elm
c : { r | b : String } -> { r | b : String }
c r = {r | b = "bla"}

withD = { b = "asdf", d = "asdf" }

d = (c withD).d
```

No error about missing a field `d`, with the same return type of `c` being `{ r | b : String }`.

**Comment by mgold** [comment]
In these two examples, function `c` is defined the same way, but the type variable `r` is inferred differently. In the first example, `a : { b : String }` so `r` is inferred to be `{}`, the empty record, because there are no extra fields. In the second, `withD : { b : String, d : String }` so `r` is inferred to be `{ d : String }`. The compiler correctly detects determines which fields are present after the call to the polymorphic function, taking the original record into account. If you're still confused, try writing the signatures of `c a` with `c withD`.

In short, I agree with Janis that the only bug here is the ever-persistent #873, and perhaps #1329. Nothing new.

**Comment by mgold** [comment]
To be fair,

>  It's absolutely okay to access arbitrary fields if the value was declared to have arbitrary fields

is better stated as, "it's absolutely okay to allow a record to have arbitrary fields in addition to the ones you are using, which will appear in the type signature".

**Comment by jvoigtlaender** [comment]
Max, the type declaration
```elm
foo : { b | x : String }
```
does *exactly* say that `foo` *has* (not only *is allowed to have*) all possible fields of all possible types.

**Comment by FPtje** [comment]
I think I get it. That type is invalid anyway. There should be no way to construct a constant with this type:

```elm
foo : { b | x : String }
```

After all, the definition `foo = {bar | x = "someString}` would need some `bar` with type `b`. Once such `bar` exists, you would have an `undefined`, which I understood is not something Elm wants.

Once one *can* construct this foo (and we can, in the numerous ways shown before), it can have any member, because it is polymorphic.

So once that infinite type/recursion thing is fixed, it should be impossible to define `foo`. At that point both the problems mentioned in this issue are resolved.

**Comment by mgold** [comment]
In the context of a record value, this makes sense. I agree that I should not be able to annotate a constant with a `{ b |`-style annotation.

But for functions like `c` from above, I still think it means *allowed to have* arbitrary fields of arbitrary types that do not conflict with those specified. l can call `c { b = "bar" }` just fine, despite passing it a record that does *not* contain all possible fields with all possible types.

This notion of "allowed to have" doesn't make sense for a literal constant, so it's forbidden.

**Comment by jvoigtlaender** [comment]
@FPtje, you got it exactly right.

@mgold, you are incoherent. Saying "But for functions like `c` above, I still think ..." is strange when the formulation you were suggesting is not correct was *not at all* about functions like `c`. I'm not sure you understand polymorphism correctly. You might want to think about covariant and contravariant occurrences of type variables to get what the fundamental difference is between the roles of the `b` in
```elm
f : { b | x : String }
```
and the role of `b` in
```elm
g : { b | x : String } -> ...
```
It's not that the notion of  "allowed to have" is forbidden for `f`, it's that it's not applicable for `f`, since the `b` there has a meaning different from "I'm allowed to have ...".

**Comment by jvoigtlaender** [comment]
In the hope of making this yet clearer, think about this declaration:
```elm
list : List a
```
It's not saying that `list` is "allowed" to have any element type. It says that `list` "has" every element type. For example, this usage will be valid and type-correct:
```elm
test = list ++ [ list, [ pi ] ]
```
and in there, `list` has both the type `List Float` (otherwise `[ list, [ pi ] ]` wouldn't make sense) and the type `List (List Float)` (otherwise the `list ++` there wouldn't make sense).

And there's nothing that can be "forbidden" here. An implementation of
```elm
list : List a
```
does exist in Elm, without nontermination or `Debug.crash` trickery.
